lib/src/test/java/tech/relaycorp/awaladroid/messaging/IncomingMessageTest.kt

```diff
@@ -32,11 +62,168 @@ internal class IncomingMessageTest : MockContextTestCase() {
+                "Ignoring PDA path from public endpoint ${thirdPartyEndpoint.privateAddress} " +
+                    "(${thirdPartyEndpoint.publicAddress})"
+            )
+        )
+    }
+
+    @Test
+    fun build_pdaPath_malformed() = runBlockingTest {
+        val channel = createEndpointChannel(RecipientAddressType.PRIVATE)
+        val parcel = Parcel(
+            channel.firstPartyEndpoint.privateAddress,
+            makePDAPathParcelPayload(channel, "malformed".toByteArray()),
+            PDACertPath.PDA,
+        )
+        val ack = StubACK()
+
+        val message = IncomingMessage.build(parcel, ack::run)
+
+        assertNull(message)
+        assertTrue(ack.wasCalled)
+        verify(storage.privateThirdParty, never()).set(any(), any())
+        assertTrue(
+            logCaptor.infoLogs.contains(
+                "Ignoring malformed PDA path for ${channel.firstPartyEndpoint.privateAddress} " +
+                    "from ${channel.thirdPartyEndpoint.privateAddress}"
+            )
+        )
+    }
+
+    @Test
+    fun build_pdaPath_invalid() = runBlockingTest {
+        val channel = createEndpointChannel(RecipientAddressType.PRIVATE)
+        val now = ZonedDateTime.now()
+        val expiredPDA = issueDeliveryAuthorization(
+            channel.firstPartyEndpoint.publicKey,
+            KeyPairSet.PDA_GRANTEE.private,
+            now.minusSeconds(1),
+            thirdPartyEndpointCertificate,
+            now.minusSeconds(2),
+        )
+        val parcel = Parcel(
+            channel.firstPartyEndpoint.privateAddress,
+            makePDAPathParcelPayload(
+                channel,
+                CertificationPath(expiredPDA, listOf(thirdPartyEndpointCertificate))
+            ),
+            PDACertPath.PDA,
+        )
+        val ack = StubACK()
+
+        val message = IncomingMessage.build(parcel, ack::run)
+
+        assertNull(message)
+        assertTrue(ack.wasCalled)
+        verify(storage.privateThirdParty, never()).set(any(), any())
+        assertTrue(
+            logCaptor.infoLogs.contains(
+                "Ignoring invalid PDA path for ${channel.firstPartyEndpoint.privateAddress} " +
+                    "from ${channel.thirdPartyEndpoint.privateAddress}"
+            )
+        )
+    }
+
+    @Test
+    fun build_pdaPath_valid() = runBlockingTest {
+        val channel = createEndpointChannel(RecipientAddressType.PRIVATE)
+        val pda = issueDeliveryAuthorization(
+            channel.firstPartyEndpoint.publicKey,
+            KeyPairSet.PDA_GRANTEE.private,
+            thirdPartyEndpointCertificate.expiryDate,
+            thirdPartyEndpointCertificate,
+        )
+        val pdaPath = CertificationPath(pda, listOf(thirdPartyEndpointCertificate))
+        val parcel = Parcel(
+            channel.firstPartyEndpoint.privateAddress,
+            makePDAPathParcelPayload(channel, pdaPath),
+            PDACertPath.PDA,
+        )
+        val ack = StubACK()
```

Comment:
I only defined `StubACK` because I couldn't find something like `val ack = spy<suspend () -> Unit>()` to work with `verify(ack)()`: I got an error about `ack` being invoked with a different `Continuation` argument, which I guess means this approach doesn't work with coroutines.

I also tried `verify(ack)(any())` but then that failed because the function doesn't take any arguments. So I guess the `Continuation` instance that gets passed as an argument is just an internal thing in Kotlin coroutines.

So, does JUnit or Mockito offer a way to stub a standalone function with no side-effects? I searched online but couldn't find a way, so I created `StubACK`.

Reply:
I imagine `mock<suspend () -> Unit>()` didn't work as well?

The StubACK sounds like a decent alternative. In other projects we've been trying out https://mockk.io and it feels to have better answers for some of this kotlin coroutines challenges.

Reply:
I _think_ I tried `mock()` but I'm not entirely sure, so I'll have another go.

Mockk looks pretty good, though it'd probably be best to replace Mockito altogether instead of keeping both.

Reply:
Yeah, I get the same error with `mock()`:

```
Argument(s) are different! Wanted:
function1.invoke(
    Continuation at tech.relaycorp.awaladroid.messaging.IncomingMessageTest$build_pdaPath_fromPublicEndpoint$1.invokeSuspend(IncomingMessageTest.kt:123)
);
-> at tech.relaycorp.awaladroid.messaging.IncomingMessageTest$build_pdaPath_fromPublicEndpoint$1.invokeSuspend(IncomingMessageTest.kt:123)
Actual invocations have different arguments:
function1.invoke(
    Continuation at tech.relaycorp.awaladroid.messaging.IncomingMessage$Companion.build$lib_debug(IncomingMessage.kt:68)
);
-> at tech.relaycorp.awaladroid.messaging.IncomingMessage$Companion.build$lib_debug(IncomingMessage.kt:68)

Argument(s) are different! Wanted:
function1.invoke(
    Continuation at tech.relaycorp.awaladroid.messaging.IncomingMessageTest$build_pdaPath_fromPublicEndpoint$1.invokeSuspend(IncomingMessageTest.kt:123)
);
-> at tech.relaycorp.awaladroid.messaging.IncomingMessageTest$build_pdaPath_fromPublicEndpoint$1.invokeSuspend(IncomingMessageTest.kt:123)
Actual invocations have different arguments:
function1.invoke(
    Continuation at tech.relaycorp.awaladroid.messaging.IncomingMessage$Companion.build$lib_debug(IncomingMessage.kt:68)
);
-> at tech.relaycorp.awaladroid.messaging.IncomingMessage$Companion.build$lib_debug(IncomingMessage.kt:68)
```

So I'll leave `StubACK`.

Reply:
Here's how I used it btw:

```kotlin
    @Test
    fun build_pdaPath_fromPublicEndpoint() = runBlockingTest {
        val channel = createEndpointChannel(RecipientAddressType.PUBLIC)
        val parcel = Parcel(
            channel.firstPartyEndpoint.privateAddress,
            makePDAPathParcelPayload(channel, "doesn't matter".toByteArray()),
            PDACertPath.PDA,
        )
        val ack = mock<suspend () -> Unit>()

        val message = IncomingMessage.build(parcel, ack)

        assertNull(message)
        verify(ack)()
        val thirdPartyEndpoint = channel.thirdPartyEndpoint as PublicThirdPartyEndpoint
        assertTrue(
            logCaptor.infoLogs.contains(
                "Ignoring PDA path from public endpoint ${thirdPartyEndpoint.privateAddress} " +
                    "(${thirdPartyEndpoint.publicAddress})"
            )
        )
    }
```